mgr/dashboard_v2: Configuration settings support

rjfd opened the pull request:
This PR adds the infrastructure to manage configuration settings in an easy way. For more information check the additions to the HACKING.rst file.

The RGW management feature depends on this feature.

(This code was already reviewed in https://github.com/openattic/ceph/pull/46)

Signed-off-by: Ricardo Dias rdias@suse.com

votdev commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -463,3 +463,48 @@ The unit test code will look like the following::
+  $ ceph dashboard get-admin-email-address
+  $ ceph dashboard set-admin-email-address <value>
+
+To access, or modify the config setting value from your python code, either
```
s/python/Python/
Plese check the whole file for further events.

p-na commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -463,3 +463,48 @@ The unit test code will look like the following::
+
+    ADMIN_EMAIL_ADDRESS = ('admin@admin.com', str)
+
+The value of the class attribute is pair composed by the default value for that
```
I suppose `is pair` should be `is a pair`.

jcsp commented on src/pybind/mgr/dashboard_v2/settings.py:
```diff
@@ -0,0 +1,113 @@
+
+def handle_option_command(cmd):
+    if cmd['prefix'] not in _OPTIONS_COMMAND_MAP:
+        return (-errno.EINVAL, '', "Command not found '{}'".format(cmd['prefix']))
```
I'd use ENOSYS to indicate that this function can't handle the command (but it may be a valid command for someone else).  Then when calling this function, instead of checking r == 0 to decide if it was handled, you can use r != errno.ENOSYS

rjfd replied:
will do

callithea commented on src/pybind/mgr/dashboard_v2/HACKING.rst:
```diff
@@ -463,3 +463,48 @@ The unit test code will look like the following::
+The value of the class attribute is pair composed by the default value for that
+setting, and the python type of the value.
+
+By declaring the ``ADMIN_EMAIL_ADDRESS`` class attribute, when you restart the
```
Suggestion: "By declaring [...] and restarting the dashboard plugin [...]"

jcsp commented on src/pybind/mgr/dashboard_v2/settings.py:
```diff
@@ -0,0 +1,113 @@
+    GRAFANA_API_SCHEME = ('http', str)
+
+    # RGW settings
+    RGW_API_HOST = ('', str)
```
I think last time we were looking at similar code, there was some discussion of getting the RGW settings automatically so that not manual config was needed -- anybody been working in that area?  @mattbenjamin ?

Liuchang0812 replied:
@jcsp do you mean that RGW daemons report those configs (HOST, PORT ...) to ceph-mgr as daemon's metadata?

p-na replied:
This setting will be used by the RGW proxy (in the works, no PR yet) to override the automatic detection of a host and port. The RGW proxy will use the service map to get the host and port, but I'm not 100% sure that this is correct for most setups, but, AFAIK, it's not possible to get more data as a Ceph MGR module (like the data available through `radosgw-admin`).

rjfd replied:
Actually, I'm going to remove all the options from this PR. These options should be added by the PR that needs them. This PR should only add the infrastructure.

tchaikov commented on src/pybind/mgr/dashboard_v2/settings.py:
```diff
@@ -0,0 +1,113 @@
+
+# pylint: disable=no-init
+class Settings(with_metaclass(SettingsMeta, object)):
+    pass
```
@rjfd https://github.com/ceph/ceph-ci/commit/082c0b12eb75a6fe3b2b4f783fb4aa7c00507857

sebastian-philipp replied:
I had a similar issue and solved it by using [`@add_metaclass`](https://pythonhosted.org/six/#six.add_metaclass)

sebastian-philipp replied:
What I meant was: I got that exception, too. I fixed that locally by using `@add_metaclass` instead.

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/settings.py:
```diff
@@ -0,0 +1,113 @@
+
+class SettingsMeta(type):
+
+    def __getattribute__(cls, attr):
```
Is there a reason, you're not using `__getattr__` here?

rjfd commented:
@tchaikov @sebastian-philipp addressed your comments. Fixed the metaclass problem with @sebastian-philipp suggestion.

Ran the dashboard_v2 suite successfully:
http://pulpito.ceph.com/rdias-2018-03-07_11:48:14-rados:mgr-wip-rdias-testing-distro-basic-mira/

rjfd commented:
@jcsp your comments were addressed as well

rjfd commented:
@p-na @votdev addressed your comments

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/settings.py:
```diff
@@ -0,0 +1,104 @@
+
+class SettingsMeta(type):
+    def __getattr__(cls, attr):
+        if hasattr(Options, attr):
```
`hasattr` looks like this:
```python
def __hasattr__(self, name):
    try:
        getattr(self, name)
        return True
    except AttributeError: 
        return False
```
No need to duplicate catching `AttributError` here. What about simplifying `SettingsMeta.__getattr__` to:
```python
def __getattr__(cls, attr):
    default, stype = getattr(Options, attr)
    return stype(mgr.get_config(attr, default))
```
?

rjfd replied:
sounds good

rjfd commented:
@sebastian-philipp changed `__getattr__` accordingly to your suggestion.

rjfd commented:
@tchaikov did you include the latest changes of this PR in your testing branch? or should I test it?

tchaikov commented:
@rjfd yeah, i pulled your branch after you pushed the latest version.

tchaikov commented:
http://pulpito.ceph.com/kchai-2018-03-08_12:59:37-rados-wip-kefu-testing-2018-03-08-1932-distro-basic-smithi/

- the swift failure was caused by #20419
- and i reran `osd-pool-create.sh` multiple times locally, and was not able to reproduce that failure.
